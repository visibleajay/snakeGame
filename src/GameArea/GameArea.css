
.GameArea {
    position: relative;
    width: 600px;
    height: 400px;
    border: 1px solid green;
}

.GameOver {
    display: flex;
    width: 100%;
    height: 100%;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    background-color: #c3c3c3;
    opacity: 0.25;
    color: black;
    font-size: 4em;
}

