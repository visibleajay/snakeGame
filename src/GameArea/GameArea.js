
import React from 'react';
import Point from '../Point/Point.js';
import Snake from '../Snake/Snake.js';

import { connect } from 'react-redux';

import "./GameArea.css"

class GameArea extends React.Component {

    constructor(props) {
        super(props);
        // const foodPos       =   { "x": 200, "y": 140 };
        // const isGameOver    =   false;
        // const leftKey       =   37;
        // this.state = {foodPos, isGameOver, "gameBoundary": {}, "keyCode": leftKey};
        this.handleSnakeMovement    =   this.handleSnakeMovement.bind(this);
    }

    gameAreaRefCallback     =   (element) => {
        if ( element ) {
            element.focus();
            let { left, right, top, bottom } = element.getBoundingClientRect();
            const gameBoundary  =   {
                    "left": 0, "right": right - left,
                    "top": 0, "bottom": bottom - top
            };
            const action = {"type": "Add_Game_Boundary", gameBoundary };
            this.props.dispatch(action);
            // this.setState({
            //     gameBoundary
            // });
        }
    }

    handleSnakeMovement(event) {

        const keyCode = event.keyCode;

        const possibleKeyCodes = [37, 38, 39, 40];

        const isSimilarOrParallelKeyPressed = (code) => code%2 === this.props.keyCode%2;

        if ( !possibleKeyCodes.includes(keyCode) || isSimilarOrParallelKeyPressed(keyCode) ) {
            return ;   
        }

        const action = { "type": "Update_Key_Code", keyCode };
        this.props.dispatch(action);
        // this.setState({
        //     keyCode
        // });
    }

    handleGameOver = (isGameOver) => {
        const action = { "type": "Stop_Game", isGameOver };
        this.props.dispatch(action);
        // this.setState({isGameOver});
    }

    handleFoodEat  = (snakePositions) => {
        let foodPos = null;
        
        const isUniquePos = ( {x: foodPosX, y: foodPosY}) => snakePositions.some( pos => pos.x === foodPosX && pos.y === foodPosY );

        do {
            const {right, bottom}   =   this.props.gameBoundary;

            let randomLeft   = Math.floor(Math.random()*right);
            let randomBottom = Math.floor(Math.random()*bottom);

            if ( randomLeft >= right ) randomLeft = right - 20;

            if ( randomBottom >= bottom ) randomBottom = bottom - 20;

            foodPos = { x: randomLeft - randomLeft%20, y: randomBottom - randomBottom%20 };    
        } while ( isUniquePos( foodPos ))

        const action    =   {"type": "Update_Food_Position" , foodPos};
        this.props.dispatch(action);
        // this.setState({foodPos});
    }

    render() {
        return (
            <div ref={this.gameAreaRefCallback} className="GameArea" onKeyUp={this.handleSnakeMovement} tabIndex="0">
                <Snake foodPos={this.props.foodPos} 
                       gameBoundary={this.props.gameBoundary} 
                       keyCode={this.props.keyCode} 
                       onGameOver={(isGameOver) => this.handleGameOver(isGameOver)}
                       onFoodEat={(snakePositions) => this.handleFoodEat(snakePositions)}/>
                <Point xPos={this.props.foodPos.x} yPos={this.props.foodPos.y} />
                {this.props.isGameOver ? <div className="GameOver">Game Over</div> : ""}
            </div>
        )
    }
}

const mapStateToProps = state => (
    { ...state }
);

export default connect(mapStateToProps)(GameArea);
