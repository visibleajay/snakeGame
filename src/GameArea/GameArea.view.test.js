
import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';

import "@testing-library/jest-dom/extend-expect";

import GameArea from './GameArea.js';
import advanceTimeBy from '../shared.js';

describe("<GameArea />", () => {

    beforeAll( () => {
        jest.useFakeTimers();
    });

    afterEach(cleanup);

    it( "Initial focus should be on GameArea", () => {
        const { container } = render(<GameArea />);
        // 
        expect(document.activeElement).toEqual(container.firstElementChild);
    });

    it( "Press down key and move the snake in down direction", () => {
        const { container, getAllByRole } = render(<GameArea />);

        advanceTimeBy(150);

        const [firstSnakePosition, secondSnakePosition] = getAllByRole("Box");
        expect(firstSnakePosition).toHaveStyle(`top: 100px; left: 180px;`);
        expect(secondSnakePosition).toHaveStyle(`top: 100px; left: 200px;`);

        // keyBoard event issue in triggering
        fireEvent.keyUp(container.firstChild, {key: "ArrowDown", keyCode: 40, code: 40});
        advanceTimeBy(150);
        expect(getAllByRole("Box")).toMatchSnapshot();
    });
});

/*

1. Active Element is container.
2. Check direction of snake, in which it is moving.
3. On Key down press, the snake should move downwards.
4. On key up press, the snake should move downwards.
5. After specific seconds, Game Over should be visible.
6. On space-bar, the game should restart again and snake should move in left direction from inital state.
*/

