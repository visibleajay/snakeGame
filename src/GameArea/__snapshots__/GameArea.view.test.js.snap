// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameArea /> Press down key and move the snake in down direction 1`] = `
Array [
  <div
    class="Box"
    role="Box"
    style="top: 100px; left: 180px;"
  />,
  <div
    class="Box"
    role="Box"
    style="top: 100px; left: 200px;"
  />,
]
`;

exports[`GameArea Comp Press down key and move the snake in down direction 1`] = `
Array [
  <div
    class="Box"
    role="Box"
    style="top: 100px; left: 180px;"
  />,
  <div
    class="Box"
    role="Box"
    style="top: 100px; left: 200px;"
  />,
]
`;
